### DISCUSSION TITLE
Using GAT to implement node classification of bipartite graphs, how do model messages pass?

### DISCUSSION THREAD

**Discussion by Zhang-1218** [pyg-team/pytorch_geometric#4159] [Q&A]
Hi, I've been trying to do node classification in a bipartite graph using pyg recently, and I'm running into some confusion along the way.
My understanding is that the data is an undirected bipartite graph, so message passing is performed in both directions on all edges, right?
See the schematic diagram of message passing in the heterogeneous graph given on the official website [https://pytorch-geometric.readthedocs.io/en/latest/notes/heterogeneous.html](url)
, but I'm a little confused about the message passing in an undirected bipartite graph.

The messages passed in the undirected bipartite graph I draw are shown below. 
![image](https://user-images.githubusercontent.com/28587429/155846541-9326184a-3fc1-46be-95a1-d317ab4bb617.png)

Is my understanding correct?

Taking paper-author data as an example, a bipartite graph is formed to classify papers or authors. Is messaging as pictured above?
In actual training, if it is to predict the category of the paper, then it is to use the label of the paper to calculate the network loss for learning, and finally output the category of the paper.
If it is to predict the author's class, use the author's label to calculate the network loss for learning, right?
Looking forward to your answer。

**Comment by rusty1s** [comment]
This looks mostly correct, although the inputs of a convolution operator receive two inputs (two arrows), and just produce a single output (one arrow), as shown in the original figure. That is, source node features send messages to destination node, both feature representations get combined and finally produce a new output destination node feature. Hope this clarifies your issues.

**Reply by Zhang-1218** [reply]
thank you for your reply
The above diagram is based on the official website and contains two inputs and two outputs, but this is not the case.

Suppose the data is an undirected bipartite graph, containing two types of nodes
node_type1=src=author, feature dimension is 10.
node_type2=dst=paper, feature dimension is 5.
The author-write-paper (paper-rev_write-author) edge has attributes and dimension is 3.

Predict the category of a paper by GAT. Although the feature dimensions of the two types of nodes are not the same, they do not affect the aggregation of information between nodes. And this aggregation relationship is bidirectional, that is, from src->dst, but also from dst->src. Is my understanding correct?

At the same time, I drew the following two message passing diagrams, including diagram A and diagram B, which one is correct?
![image](https://user-images.githubusercontent.com/28587429/155874878-b1491364-4360-45fb-be44-0a466583fa92.png)

**Reply by rusty1s** [reply]
Yes, your understanding is totally correct, but the figures still have wrong arrows. That is, both paper and author should have an arrow to both `GATConv` layers, and each `GATConv` will produce a single output arrow. In addition, only the output arrow of the paper "class" after the last GNN layer should be maintained (right figure).

**Reply by Zhang-1218** [reply]
Emmm, thanks for your reply.
I think I roughly understand the message passing process

The GAT network model is shown in the following code
```
class GAT(torch.nn.Module):
    def __init__(self, hidden_channels, out_channels):
        super().__init__()
        self.conv1 = GATConv((-1, -1), hidden_channels, add_self_loops=False)
        self.lin1 = Linear(-1, hidden_channels)
        self.conv2 = GATConv((-1, -1), out_channels, add_self_loops=False)
        self.lin2 = Linear(-1, out_channels)

    def forward(self, x, edge_index):
        x = self.conv1(x, edge_index) + self.lin1(x)
        x = x.relu()
        x = self.conv2(x, edge_index) + self.lin2(x)
        return x

```

Therefore, Figure A and Figure B are actually the same in essence, except that the convolution of two different nodes is displayed separately in Figure B. What should the figure look like?

![image](https://user-images.githubusercontent.com/28587429/155876227-cef36d52-78f4-45df-9a36-35d9158a9730.png)

**Reply by rusty1s** [reply]
Yes, Figure B looks mostly correct now. Figure A misses the concept of having separate weights/GNN operators for individual edge types. Importantly, I think that only the "paper" input arrow to the last Relu should be maintained, the other one needs to be dropped.

**Comment by Zhang-1218** [comment]
Emmm
So, the final figure looks like this?

![image](https://user-images.githubusercontent.com/28587429/155876861-f3737f32-7d52-496b-a82d-2b9d9621a004.png)

**Reply by rusty1s** [reply]
Yes, this looks correct. It seems that you also dropped the second `GATConv` (since its output is not used).

**Reply by Zhang-1218** [reply]
Yes, because the "author" also completes the second layer GATConv though. But didn't apply the result it produced, so I removed it to make it clearer. Do you think it is okay?

**Reply by rusty1s** [reply]
Yes, this looks good to me now.

**Reply by Zhang-1218** [reply]
Emm, I already understand.
thank you so much for your patience and answers! 
have a pleasant workday!

**Reply by Angelo92Git** [reply]
Hi is this consistent with the documentation?
![hetero](https://github.com/pyg-team/pytorch_geometric/assets/52296518/2f80f096-6946-451d-be32-db34de8da166)
